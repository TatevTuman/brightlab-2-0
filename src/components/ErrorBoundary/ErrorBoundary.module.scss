.error-boundary {
  color: var(--danger)
}
