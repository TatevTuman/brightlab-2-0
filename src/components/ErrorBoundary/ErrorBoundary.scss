.error-boundary {
  flex: 1;

  &-title {

  }

  &-description {
    margin-bottom: 1em;
  }

  &-hint {

  }

  &-graphQL {
    overflow-x: scroll;
  }
}

.error {
  &-name {

  }

  &-message {
    margin-bottom: 1em;
  }

  &-info {
    color: var(--danger);
    margin-bottom: 2em;
  }
}
