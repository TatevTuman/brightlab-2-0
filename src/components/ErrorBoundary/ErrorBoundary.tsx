import React, { ErrorInfo, PureComponent, ReactNode } from 'react'
import { GraphQLError } from '@types'
import './ErrorBoundary.scss'

type UnionError = Error & { graphQLErrors?: GraphQLError[] }

class ErrorWithGraphQLErrors extends Error {
  graphQLErrors?: GraphQLError[]

  constructor(error: UnionError) {
    const { message, stack, graphQLErrors } = error
    super(message)

    this.message = message
    this.stack = stack
    this.graphQLErrors = graphQLErrors
  }
}

class LoadContentError extends Error {
  constructor(error: string) {
    super(error)
    this.name = 'LoadContentError'
  }
}

class LoadQueryContentError extends ErrorWithGraphQLErrors {
  constructor(error: UnionError) {
    super(error)
    this.name = 'LoadQueryContentError'
  }
}

class LoadMutationContentError extends ErrorWithGraphQLErrors {
  constructor(error: UnionError) {
    super(error)
    this.name = 'LoadMutationContentError'
  }
}

interface ErrorBoundaryProps {
  title?: string
  description?: string
  hint?: string
}

interface ErrorBoundaryState {
  error: UnionError | null
  errorInfo: string
  hasError: boolean
}

class ErrorBoundary extends PureComponent<ErrorBoundaryProps, ErrorBoundaryState> {
  state: ErrorBoundaryState = {
    error: null,
    errorInfo: '',
    hasError: false
  }

  static getDerivedStateFromError(error: UnionError): Partial<ErrorBoundaryState> {
    return {
      error,
      hasError: true
    }
  }

  componentDidCatch(error: UnionError, errorInfo: ErrorInfo): void {
    this.setState({
      errorInfo: errorInfo.componentStack
    })
  }

  renderErrorDefault = (): JSX.Element => {
    const { error, errorInfo } = this.state

    return (
      <>
        <h2 className={'p__primary_sm error-name'}>{error?.name}</h2>
        <h3 className={'p__primary_sm error-message'}>{error?.message}</h3>
        <p className={'p__primary_sm error-info'}>{errorInfo}</p>
      </>
    )
  }

  renderErrorProps = (): JSX.Element => {
    const { title, description, hint } = this.props

    return (
      <>
        {title && <h2 className="error-boundary-title">Ooopps... {title}</h2>}
        {description && <h3 className="error-boundary-description">{description}</h3>}
        {hint && <p className="p__secondary-sm error-boundary-hint">{hint}</p>}
      </>
    )
  }

  renderGraphQLErrors = (graphQLErrors?: GraphQLError[]): JSX.Element | null => {
    if (!graphQLErrors || !graphQLErrors.length) return null

    return (
      <div className="error-boundary-graphQL">
        <h2>GraphQl errors</h2>
        <pre>
          {graphQLErrors.map((graphQLError, index) => {
            const { locations, message } = graphQLError

            return <div key={index}>{message}</div>
          })}
        </pre>
      </div>
    )
  }

  render(): ReactNode {
    const { error, hasError } = this.state

    if (hasError) {
      return (
        <section className="error-boundary">
          {this.renderErrorProps()}
          {this.renderErrorDefault()}
          {this.renderGraphQLErrors(error?.graphQLErrors)}
        </section>
      )
    }

    return this.props.children
  }
}

export default ErrorBoundary
export { ErrorWithGraphQLErrors, LoadContentError, LoadQueryContentError, LoadMutationContentError }
