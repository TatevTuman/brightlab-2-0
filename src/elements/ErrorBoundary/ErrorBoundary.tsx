import React, { PureComponent } from 'react'
import './ErrorBoundary.scss'

class ErrorBoundary extends PureComponent {
  static getDerivedStateFromError(error: Error) {
    console.log('error', error)
    return { hasError: true }
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.log('error', error)
    console.log('errorInfo', errorInfo)
  }

  render() {
    return this.props.children
  }
}

export default ErrorBoundary
