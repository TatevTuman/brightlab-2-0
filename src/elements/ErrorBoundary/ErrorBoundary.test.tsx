import React from 'react'
import renderer from 'react-test-renderer'
import ErrorBoundary from './ErrorBoundary'

describe('ErrorBoundary', () => {
  it('render()', () => {
    const errorBoundary = renderer.create(<ErrorBoundary>Test</ErrorBoundary>).toJSON()
    expect(errorBoundary).toMatchSnapshot()
  })
})
