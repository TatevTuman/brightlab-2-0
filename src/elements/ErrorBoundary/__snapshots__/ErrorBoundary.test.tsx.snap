// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary render() 1`] = `"Test"`;
